### PR TITLE
AEJ-1605:Check contents within input structures are valid before use

### PR DESCRIPTION
**Pull request by mkuligow**

### CODE REVIEW

File: src/main/java/com/intel/gkl/compression/IntelDeflater.java [hunk 1]
```diff
@@ -169,6 +169,12 @@ public void finish() {
      */
     @Override
     public int deflate(byte[] b, int off, int len ) {
```
Comment (SnehalA) [inline src/main/java/com/intel/gkl/compression/IntelDeflater.java:171]: Can you please check if we need `off` ?

Comment (Kmannth) [reply src/main/java/com/intel/gkl/compression/IntelDeflater.java:171]: https://github.com/Intel-HLS/GKL/blob/master/src/main/java/com/intel/gkl/compression/IntelDeflater.java  has the API.  

Off is offset from the byte pointer.  It seems normal type thing to me.   Need to be sure off + len is not greater then sizeof b.

Comment (SnehalA) [reply src/main/java/com/intel/gkl/compression/IntelDeflater.java:171]: I am trying to understand where it's assigned. It should be either [declared](https://github.com/Intel-HLS/GKL/blob/master/src/main/java/com/intel/gkl/compression/IntelDeflater.java#L121) and passed with java object as inputBufferOffset in the structure, however it is commneted in counterpart
[here](https://github.com/Intel-HLS/GKL/blob/08747449e8fc9d42e4ac270ef5077ada9d19ed0a/src/main/native/compression/IntelDeflater.cc#L198)

File: src/main/java/com/intel/gkl/smithwaterman/IntelSmithWaterman.java [hunk 2]
```diff
@@ -103,7 +110,7 @@ public SWNativeAlignerResult align(byte[] refArray, byte[] altArray, SWParameter
         return new SWNativeAlignerResult(new String(cigar).trim(), offset);
     }
 
-    public int getStrategy(SWOverhangStrategy strategy)
+    public int getStrategy(SWOverhangStrategy strategy) //TODO: Check with GKL Team
```
Comment (SnehalA) [inline src/main/java/com/intel/gkl/smithwaterman/IntelSmithWaterman.java:113]: We can add null check for strategy here

File: src/main/java/com/intel/gkl/compression/IntelDeflater.java [hunk 3]
```diff
@@ -169,6 +169,12 @@ public void finish() {
...
+        if(b == null) {
+            throw new NullPointerException("Input is null");
+        }
+        if(len <= 0) {
+            throw new NullPointerException("Input buffer length is less or equal zero.");
+        }
 
         return deflateNative(b, len);
```
Comment (SnehalA) [inline src/main/java/com/intel/gkl/compression/IntelDeflater.java:179]: `off` not passed to native function

File: src/main/java/com/intel/gkl/smithwaterman/IntelSmithWaterman.java (hunk 2 above)
Comment (mkuligow) [reply src/main/java/com/intel/gkl/smithwaterman/IntelSmithWaterman.java:113]: I thought about default strategy, do we have strategy with id 0? In GetStrategy function is initial value for  intStrategy.

File: src/main/java/com/intel/gkl/compression/IntelDeflater.java (hunk 3 above)
Comment (mkuligow) [reply src/main/java/com/intel/gkl/compression/IntelDeflater.java:179]: Should be?

Comment (SnehalA) [reply src/main/java/com/intel/gkl/compression/IntelDeflater.java:182]: Just calling out for disscusion on L171.

File: src/main/java/com/intel/gkl/smithwaterman/IntelSmithWaterman.java (hunk 2 above)
Comment (Kmannth) [reply src/main/java/com/intel/gkl/smithwaterman/IntelSmithWaterman.java:113]: Reviewing the code it seems getStraegy should be a private function. I did not find it in GATK anywhere and it is only use in in this class.   GKL seems to remap the Overhang Strategy type for some reason. 

In SWNativeAlignerResult we need to check either overhangStrategy or the modified  intStrategy to be a valid input before we call  alignNative. 

Do not call  alignNative if it is not a valid strategy and return an error/exception.

**Review by SnehalA — APPROVED** [review]

File: src/main/java/com/intel/gkl/compression/IntelDeflater.java (hunk 1 above)
Comment (swepnadoss) [reply src/main/java/com/intel/gkl/compression/IntelDeflater.java:174]: Suggestion to cover this validation with `if(off < 0 || off > b.length - len) `

Comment (SnehalA) [reply src/main/java/com/intel/gkl/compression/IntelDeflater.java:174]: Here `off` is dummy input can simply check for null and assign as 0; than throwing exception.

File: src/main/java/com/intel/gkl/compression/IntelDeflater.java [hunk 4]
```diff
@@ -139,9 +139,12 @@ public void setInput(byte[] b, int off, int len) throws NullPointerException {
             throw new NullPointerException("Input is null");
         }
         if(len <= 0) {
-            throw new NullPointerException("Input buffer length is zero.");
+            throw new NullPointerException("Input buffer length is less or equal zero.");
+        }
+        if(off < 0 || off > len )
+        {
+            throw new IndexOutOfBoundsException("Offset cannot be less then 0 and greater then length.");
```
Comment (SnehalA) [inline src/main/java/com/intel/gkl/compression/IntelDeflater.java:146]: Check if IllegalArgumentException makes sense for consistency, for chekcing expected value ranges for `off` and `level`

Comment (Kmannth) [reply src/main/java/com/intel/gkl/compression/IntelDeflater.java:146]: I think the function declaration will need a new "throws"  update to include IndexOutOfBoundsException or IndexOutOfBoundsException is added. 

Can you confirm?

File: src/main/java/com/intel/gkl/compression/IntelDeflater.java [hunk 5]
```diff
@@ -189,7 +192,6 @@ public int deflate(byte[] b, int off, int len ) {
     public boolean finished() {
         return finished;
     }
-
```
Comment (Kmannth) [inline src/main/java/com/intel/gkl/compression/IntelDeflater.java:186]: No needed?

File: src/main/java/com/intel/gkl/compression/IntelDeflater.java (hunk 4 above)
Comment (mkuligow) [reply src/main/java/com/intel/gkl/compression/IntelDeflater.java:146]: Hi, I added throws declaration, @Kmannth  should i change IndexOutOfBoundsException and  NullPointerException from line 142 to IllegalArgumentException?

File: src/main/java/com/intel/gkl/smithwaterman/IntelSmithWaterman.java (hunk 2 above)
Comment (SnehalA) [reply src/main/java/com/intel/gkl/smithwaterman/IntelSmithWaterman.java]: @mkuligow  Please add default to throw  IllegalArgumentException here.

File: src/main/java/com/intel/gkl/compression/IntelDeflater.java (hunk 4 above)
Comment (Kmannth) [reply src/main/java/com/intel/gkl/compression/IntelDeflater.java:146]: @mkuligow  as discussed in the call today lets move this type of error to IllegalArgumentException.

File: src/main/java/com/intel/gkl/compression/IntelDeflater.java (hunk 3 above)
Comment (Kmannth) [reply src/main/java/com/intel/gkl/compression/IntelDeflater.java:185]: As discussed to day need to check if " off != 0 "  and I would vote to print an error message and return IllegalArgumentException.   This code only works if off==0 and that is not expected so lets print a message and return without doing any work.

File: src/main/java/com/intel/gkl/smithwaterman/IntelSmithWaterman.java [hunk 6]
```diff
@@ -95,6 +95,15 @@ public synchronized boolean load(File tempDir) {
...
+            throw new NullPointerException("Reference data array is null.");
+        if(altArray == null)
+            throw new NullPointerException("Alternate data array is null.");
+        if(parameters == null)
+            throw new NullPointerException("Parameter structure is null.");
+        if(overhangStrategy == null)
+            throw new NullPointerException("OverhangStrategy is null.");
+
         int intStrategy =  getStrategy(overhangStrategy);
         byte[] cigar = new byte[2*Integer.max(refArray.length, altArray.length)];
```
Comment (SnehalA) [inline src/main/java/com/intel/gkl/smithwaterman/IntelSmithWaterman.java:108]: @mkuligow  : Please add check for cigar.length > 0, and intStrategy in range [9,12].

Comment (SnehalA) [reply src/main/java/com/intel/gkl/smithwaterman/IntelSmithWaterman.java:108]: Thank you @mkuligow Looks good!

**Review by SnehalA — APPROVED** [review]

**Review by Kmannth — APPROVED** [review]

**Pull request merged**